spec/abstract/IpcCancel_A.thy

```diff
@@ -316,28 +316,45 @@ where
     BlockedOnNotification r \<Rightarrow> Some r
   | _ \<Rightarrow> None"
 
-text \<open>
-  Sort a list of TCB refs by priority, otherwise keeping order stable.
-  In the implementation, there will only ever be one element out of order
-  or be inserted. Here, we just require a sorted result.
-\<close>
+fun
+  tcb_ep_find_index :: "obj_ref \<Rightarrow> obj_ref list \<Rightarrow> nat \<Rightarrow> (nat, 'z::state_ext) s_monad"
+where
+  "tcb_ep_find_index tptr qs curindex = do
+     prio \<leftarrow> thread_get tcb_priority tptr;
+     curprio \<leftarrow> thread_get tcb_priority (qs ! curindex);
+     if prio > curprio then
+       if curindex = 0 then return 0
+       else tcb_ep_find_index tptr qs (curindex - 1)
+     else return (curindex + 1)
+   od"
```

Comment:
A couple of things about this. One is the naming. This finds an index but quite a different one from the one `findIndex` finds. Maybe this one can have a slightly different name so that no one has a wrong expectation? Maybe something like `tcb_ep_find_prio_index`?

Another thing is, we could do without the first `thread_get` line in the definition, if we pass the `prio` as an argument? It remains constant through this function, so we don’t need to get it every time. We can get the value of `prio` before the call to this function, and pass it as an argument. Then, for the proofs, we’d want to have some projection in the preconditions that says the priority of the tcb at `tptr` is `prio`. Not sure if this helps in any way, though. But also wondering if there is any advantage of doing thread_get every time.